src/rust/cpu/cpu.rs

```diff
@@ -1816,23 +1829,35 @@ pub unsafe fn do_page_walk(
     else {
         profiler::stat_increment(TLB_MISS);
 
-        let page_dir_addr = (*cr.offset(3) as u32 >> 2).wrapping_add((page >> 10) as u32) as i32;
-        let page_dir_entry = read_aligned32(page_dir_addr as u32);
-        // XXX
-        let kernel_write_override = !user && 0 == *cr & CR0_WP;
-        if 0 == page_dir_entry & PAGE_TABLE_PRESENT_MASK {
-            // to do at this place:
-            //
-            // - set cr2 = addr (which caused the page fault)
-            // - call_interrupt_vector  with id 14, error code 0-7 (requires information if read or write)
-            // - prevent execution of the function that triggered this call
+        let pae = *cr.offset(4) & CR4_PAE != 0;
+
+        let (page_dir_addr, page_dir_entry) =
+            match walk_page_directory(pae, addr) {
+                Some((a, e)) => (a, e),
+                // to do at this place:
+                //
+                // - set cr2 = addr (which caused the page fault)
+                // - call_interrupt_vector  with id 14, error code 0-7 (requires information if read or write)
+                // - prevent execution of the function that triggered this call
+                None => return Err(PageFault {
+                    addr,
+                    for_writing,
+                    user,
+                    present: false,
+                }),
+            };
+
+        if page_dir_entry & PAGE_TABLE_PRESENT_MASK == 0 {
```

Comment:
Just in case you're wondering why this function contains so many casts: It used to be written in C where the casts were implicit. It was convert to Rust using c2rust, which made them explicit (by inserting `… as {i,u}32`). The general convention for the codebase is that virtual addresses are i32 and physical addresses u32. I will clean it up once this is merged.

Reply:
Interesting, I didn't know the project was converted from C!